truffle/src/main/java/org/jruby/truffle/Log.java

```diff
@@ -44,6 +59,59 @@ public void close() throws SecurityException {
         return logger;
     }
 
+    private static final Set<String> displayedWarnings = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    public static final String KWARGS_NOT_OPTIMIZED_YET = "keyword arguments are not yet optimized";
+
+    /**
+     * Warn about code that works but is not yet optimized as Truffle code normally would be. Only prints the warning
+     * once, and only if called from compiled code. Don't call this method from behind a boundary, as it will never
+     * print the warning because it will never be called from compiled code. Use {@link #performanceOnce} instead
+     * if you need to warn in code that is never compiled.
+     */
+    public static void notOptimizedOnce(String message) {
+        if (CompilerDirectives.inCompiledCode()) {
+            performanceOnce(message);
+        }
+    }
+
+    /**
+     * Warn about code that works but is not yet optimized as Truffle code normally would be. Only prints the warning
+     * if called from compiled code. Don't call this method from behind a boundary or transfer, as it will never print
+     * the warning because it will never be called from compiled code. Use {@link #performance} instead if you need to
+     * warn in code that is never compiled.
+     */
+    public static void notOptimized(String message) {
+        if (CompilerDirectives.inCompiledCode()) {
+            performance(message);
+        }
+    }
```

Comment:
Yes, it's trivial to add if we need it and I'd rather avoid extra methods we don't currently need.